core/service/src/cryptography/attestation/mod.rs

```diff
@@ -274,3 +265,187 @@ pub fn make_security_module(
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut f = f.debug_struct("AttestedVerifier");
+        f.finish()
+    }
+}
+
+impl AutoRefreshCertResolver {
+    pub async fn new(
+        sk: Option<Arc<PrivateSigKey>>,
+        ca_cert: Pem,
+        security_module: Arc<SecurityModuleProxy>,
+        private_vault_root_key_measurements: Option<Arc<RootKeyMeasurements>>,
+    ) -> anyhow::Result<Self> {
+        let (certified_key, expiration) = AutoRefreshCertResolver::refresh(
+            sk.clone(),
+            ca_cert.clone(),
+            security_module.clone(),
+            private_vault_root_key_measurements.clone(),
+        )
+        .await?;
+        let certified_key_with_expiration = Arc::new(RwLock::new((certified_key, expiration)));
+        let resolver = Self {
+            certified_key_with_expiration: certified_key_with_expiration.clone(),
+        };
+
+        // This task is expected to run forever as long at the node is running
+        // because AWS Nitro signing certificates expire every 24h, and
+        // ephemeral TLS certificates cannot outlive them. However, it's
+        // possible that the renewal won't succeed for a multitude of reasons:
+        //
+        // 0) bug in our code
+        // 1) bad party CA certificate
+        // 2) NSM giving up on us for reasons only known to Amazon
+        // and not signing a fresh attestation document
+        //
+        // There isn't a good way to automatically recover from any of these
+        // scenarios (definitely not from 0 or 1), and the node is not going to
+        // establish new peer connections if they happen because it won't have a
+        // valid TLS certificate anymore. Manual intervention will be required.
+        //
+        // If TLS certificates would only be generated on the node boot, we
+        // could just crash it in case of certificate generation failure, and we
+        // used to do that. Now, we regenerate certificates on a running node,
+        // and we can't simply crash it because it might already participate in
+        // many MPC sessions. We should let them finish. Also, just in case if
+        // the refresh failure is caused by a random NSM misbehaviour, we keep
+        // retrying once a minute.
+        //
+        // Still, we cannot let nodes run without valid TLS certificates. There
+        // has to be an alert if the certificate refresh failed more than 3
+        // times in a row.
+        tokio::spawn(async move {
+            loop {
+                match AutoRefreshCertResolver::refresh(
+                    sk.clone(),
+                    ca_cert.clone(),
+                    security_module.clone(),
+                    private_vault_root_key_measurements.clone(),
+                )
+                .await
+                {
+                    Ok((certified_key, expiration)) => {
+                        let mut certified_key_with_expiration =
```

Comment:
Nitpick: could you rename one of the variables? It gets very hard to follow when you override define a new variable `certified_key_with_expiration` from one with the same name